File: android_core/app/src/main/java/com/augmentos/augmentos_core/TranscriptProcessor.java
package com.augmentos.augmentos_core;

import java.util.ArrayList;
import java.util.List;

public class TranscriptProcessor {
    private int maxCharsPerLine;
    private boolean isChinese;
    
    private final int maxLines;
    private final List<String> finalTranscriptHistory = new ArrayList<>();
    private final int maxFinalTranscripts;
    private long lastPartialUpdateTime = 0;
    private static final long THROTTLE_INTERVAL_MS = 300; // 300ms throttle interval
    private final List<String> currentDisplayLines = new ArrayList<>();
    private String partialText = "";
    private String lastUserTranscript = "";

    public TranscriptProcessor(int maxCharsPerLine, int maxLines, int maxFinalTranscripts, boolean isChinese) {
        this.maxCharsPerLine = maxCharsPerLine;
        this.maxLines = maxLines;
        this.maxFinalTranscripts = maxFinalTranscripts;
        this.isChinese = isChinese;
    }

    public TranscriptProcessor(int maxCharsPerLine, int maxLines) {
        this(maxCharsPerLine, maxLines, 3, false);
    }

    public String processString(String newText, boolean isFinal) {
        if (newText == null) {
            newText = "";
        }
        newText = newText.trim();

        if (!isFinal) {
            // Store this as the current partial text (overwriting old partial)
            this.partialText = newText;
            this.lastUserTranscript = newText;

            // Combine final history with new partial text
            String combinedText = getCombinedTranscriptHistory() + " " + newText;
            currentDisplayLines.clear();
            currentDisplayLines.addAll(wrapText(combinedText, maxCharsPerLine));

            // Ensure we have exactly maxLines
            while (currentDisplayLines.size() < maxLines) {
                currentDisplayLines.add("");
            }
            while (currentDisplayLines.size() > maxLines) {
                currentDisplayLines.remove(0);
            }

            long currentTime = System.currentTimeMillis();
            long timeSinceLastUpdate = currentTime - lastPartialUpdateTime;

            // If less than throttle interval has passed, return null
            if (timeSinceLastUpdate < THROTTLE_INTERVAL_MS) {
                return null;
            }

            // Update the last update time
            lastPartialUpdateTime = currentTime;

            // Return the processed text
            return String.join("\n", currentDisplayLines);
        } else {
            // We have a final text -> clear out the partial text to avoid duplication
            this.partialText = "";

            // Add to transcript history when it's a final transcript
            addToTranscriptHistory(newText);

            // Use the same wrapping logic as partial to maintain consistency
            String combinedText = getCombinedTranscriptHistory();
            currentDisplayLines.clear();
            currentDisplayLines.addAll(wrapText(combinedText, maxCharsPerLine));

            // Ensure we have exactly maxLines
            while (currentDisplayLines.size() < maxLines) {
                currentDisplayLines.add("");
            }
            while (currentDisplayLines.size() > maxLines) {
                currentDisplayLines.remove(0);
            }

            return String.join("\n", currentDisplayLines);
        }
    }

    private void addToTranscriptHistory(String transcript) {
        if (transcript == null || transcript.trim().isEmpty()) {
            return; // Don't add empty transcripts
        }

        finalTranscriptHistory.add(transcript.trim());

        // Ensure we don't exceed maxFinalTranscripts
        while (finalTranscriptHistory.size() > maxFinalTranscripts) {
            finalTranscriptHistory.remove(0); // Remove oldest transcript
        }
    }

    private String getCombinedTranscriptHistory() {
        return String.join(" ", finalTranscriptHistory);
    }

    private List<String> wrapText(String text, int maxLength) {
        List<String> result = new ArrayList<>();
        if (text == null || text.isEmpty()) {
            return result;
        }

        if (isChinese) {
            // For Chinese text, process character by character
            StringBuilder currentLine = new StringBuilder();
            
            for (int i = 0; i < text.length(); i++) {
                char c = text.charAt(i);
                
                // If adding this character would exceed maxLength, start a new line
                if (currentLine.length() >= maxLength) {
                    result.add(currentLine.toString());
                    currentLine = new StringBuilder();
                }
                currentLine.append(c);
            }
            
            // Add the last line if it's not empty
            if (currentLine.length() > 0) {
                result.add(currentLine.toString());
            }
        } else {
            // For non-Chinese text, use word-based wrapping
            String[] words = text.split("\\s+");
            StringBuilder currentLine = new StringBuilder();

            for (String word : words) {
                // If the word itself is longer than maxLength, split it
                while (word.length() > maxLength) {
                    // Take the first maxLength characters of the word
                    String part = word.substring(0, maxLength);
                    // If we already have content in the current line, add it to result first
                    if (currentLine.length() > 0) {
                        result.add(currentLine.toString());
                        currentLine = new StringBuilder();
                    }
                    // Add this part as a new line
                    result.add(part);
                    // Keep the remaining part of the word
                    word = word.substring(maxLength);
                }
                
                // Now handle the remaining part of the word (or the whole word if it wasn't too long)
                if (currentLine.length() + word.length() + (currentLine.length() > 0 ? 1 : 0) > maxLength) {
                    // If adding this word would exceed maxLength, start a new line
                    if (currentLine.length() > 0) {
                        result.add(currentLine.toString());
                        currentLine = new StringBuilder();
                    }
                } else if (currentLine.length() > 0) {
                    // Otherwise, add a space if this isn't the first word in the line
                    currentLine.append(" ");
                }
                currentLine.append(word);
            }

            // Add the last line if it's not empty
            if (currentLine.length() > 0) {
                result.add(currentLine.toString());
            }
        }

        return result;
    }

    public void modifyLanguage(String language) {
        boolean languageIsChinese = language.equals("zh-CN");
        if (languageIsChinese != this.isChinese) {
            this.isChinese = languageIsChinese;
            if (languageIsChinese) {
                this.maxCharsPerLine = 10;
            } else {
                this.maxCharsPerLine = 30;
            }
            this.clear();
        }
    }

    public void clear() {
        finalTranscriptHistory.clear();
        currentDisplayLines.clear();
        partialText = "";
        lastPartialUpdateTime = 0;
    }
}

File: mobile/ios/Source/subs/TranscriptProcessor.swift
import Foundation

class TranscriptProcessor {
    private var maxCharsPerLine: Int
    private var maxLines: Int
    private var lines: [String]
    private var partialText: String
    private var lastUserTranscript: String
    private var finalTranscriptHistory: [String] // Array to store history of final transcripts
    private var maxFinalTranscripts: Int // Max number of final transcripts to keep
    private var isChinese: Bool
    private var currentDisplayLines: [String] // Track current display lines to maintain consistency
    private var lastPartialUpdateTime: TimeInterval = 0
    private let throttleInterval: TimeInterval = 0.3 // 300ms throttle interval
    
    init(maxCharsPerLine: Int, maxLines: Int, maxFinalTranscripts: Int = 3, isChinese: Bool = false) {
        self.maxCharsPerLine = maxCharsPerLine
        self.maxLines = maxLines
        self.lastUserTranscript = ""
        self.lines = []
        self.partialText = ""
        self.finalTranscriptHistory = []
        self.maxFinalTranscripts = maxFinalTranscripts
        self.isChinese = isChinese
        self.currentDisplayLines = []
    }
    
    func processString(_ newText: String?, isFinal: Bool) -> String? {
        let newText = newText?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        if !isFinal {
            // Store this as the current partial text (overwriting old partial)
            self.partialText = newText
            self.lastUserTranscript = newText
            
            // Combine final history with new partial text
            let combinedText = getCombinedTranscriptHistory() + " " + newText
            self.currentDisplayLines = wrapText(combinedText, maxLength: self.maxCharsPerLine)
            
            // Ensure we have exactly maxLines
            while self.currentDisplayLines.count < self.maxLines {
                self.currentDisplayLines.append("")
            }
            while self.currentDisplayLines.count > self.maxLines {
                self.currentDisplayLines.removeFirst()
            }

            let currentTime = Date().timeIntervalSince1970
            let timeSinceLastUpdate = currentTime - lastPartialUpdateTime
            
            // If less than throttle interval has passed, return nil
            if timeSinceLastUpdate < throttleInterval {
                return nil
            }
            
            // Update the last update time
            lastPartialUpdateTime = currentTime
            
            // Return the processed text
            return self.currentDisplayLines.joined(separator: "\n")
        } else {
            // We have a final text -> clear out the partial text to avoid duplication
            self.partialText = ""
            
            // Add to transcript history when it's a final transcript
            addToTranscriptHistory(newText)
            
            // Use the same wrapping logic as partial to maintain consistency
            let combinedText = getCombinedTranscriptHistory()
            self.currentDisplayLines = wrapText(combinedText, maxLength: self.maxCharsPerLine)
            
            // Ensure we have exactly maxLines
            while self.currentDisplayLines.count < self.maxLines {
                self.currentDisplayLines.append("")
            }
            while self.currentDisplayLines.count > self.maxLines {
                self.currentDisplayLines.removeFirst()
            }
            
            return self.currentDisplayLines.joined(separator: "\n")
        }
    }
    
    // Add to transcript history
    private func addToTranscriptHistory(_ transcript: String) {
        let trimmed = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return } // Don't add empty transcripts
        
        self.finalTranscriptHistory.append(trimmed)
        
        // Ensure we don't exceed maxFinalTranscripts
        while self.finalTranscriptHistory.count > self.maxFinalTranscripts {
            self.finalTranscriptHistory.removeFirst() // Remove oldest transcript
        }
    }
    
    // Get the transcript history
    func getFinalTranscriptHistory() -> [String] {
        return Array(self.finalTranscriptHistory) // Return a copy to prevent external modification
    }
    
    // Get combined transcript history as a single string
    func getCombinedTranscriptHistory() -> String {
        return self.finalTranscriptHistory.joined(separator: " ")
    }
    
    // Method to set max final transcripts
    func setMaxFinalTranscripts(_ maxFinalTranscripts: Int) {
        self.maxFinalTranscripts = maxFinalTranscripts
        // Trim history if needed after changing the limit
        while self.finalTranscriptHistory.count > self.maxFinalTranscripts {
            self.finalTranscriptHistory.removeFirst()
        }
    }
    
    // Get max final transcripts
    func getMaxFinalTranscripts() -> Int {
        return self.maxFinalTranscripts
    }
    
    private func wrapText(_ text: String, maxLength: Int) -> [String] {
        var result: [String] = []
        var remainingText = text
        
        while !remainingText.isEmpty {
            if remainingText.count <= maxLength {
                result.append(remainingText)
                break
            } else {
                var splitIndex = maxLength
                
                if isChinese {
                    // For Chinese text, find the last valid word boundary
                    // Note: For Chinese, we might want to implement a more sophisticated word boundary detection
                    // For now, we'll just split at maxLength for Chinese
                    splitIndex = maxLength
                } else {
                    // For non-Chinese text, find the last space before maxLength
                    let endIndex = remainingText.index(remainingText.startIndex, offsetBy: maxLength, limitedBy: remainingText.endIndex) ?? remainingText.endIndex
                    let searchRange = remainingText.startIndex..<endIndex
                    if let lastSpaceIndex = remainingText.rangeOfCharacter(from: .whitespaces, options: .backwards, range: searchRange)?.lowerBound {
                        splitIndex = remainingText.distance(from: remainingText.startIndex, to: lastSpaceIndex)
                    }
                    // If we didn't find a space, force split
                    if splitIndex == 0 {
                        splitIndex = maxLength
                    }
                }
                
                let splitIndexStr = remainingText.index(remainingText.startIndex, offsetBy: splitIndex)
                let chunk = String(remainingText[..<splitIndexStr]).trimmingCharacters(in: .whitespacesAndNewlines)
                if !chunk.isEmpty {
                    result.append(chunk)
                }
                remainingText = String(remainingText[splitIndexStr...]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        
        return result
    }
    
    public func getLastUserTranscript() -> String {
        return self.lastUserTranscript
    }
    
    public func clear() {
        self.lines = []
        self.partialText = ""
        self.finalTranscriptHistory = []
        self.currentDisplayLines = []
        self.lastPartialUpdateTime = 0
    }
    
    public func getMaxCharsPerLine() -> Int {
        return self.maxCharsPerLine
    }
    
    public func getMaxLines() -> Int {
        return self.maxLines
    }
}
